sql/execution_pipeline_impl.cc

```diff
@@ -2264,6 +2588,339 @@ outputScope() const
     return outputScope_;
 }
 
+/*****************************************************************************/
+/* DATASET FUNCTION ELEMENT                                                  */
+/*****************************************************************************/
+
+DatasetFunctionElement::
+DatasetFunctionElement(std::shared_ptr<PipelineElement> root,
+                       std::shared_ptr<DatasetFunctionExpression> function,
+                       GetParamInfo getParamInfo)
+ : source_(std::move(root)), function_(function)
+{
+    ExcAssert(function->args.size() <= 2);
+    ExcAssert(function->args.size() > 0);
+    ExcAssert(source_);
+    ExcAssert(getParamInfo);
+
+    if (function_->functionName == "sample") {
```

Comment:
Can we implement with virtual methods rather than hardcoding the names here?  It feels like we're coupling generic and specific logic.